xdsl/dialects/mod_arith.py

```diff
@@ -0,0 +1,74 @@
+"""
+mod_arith is a dialect implementing modular arithmetic, originally
+implemented as part of the HEIR project (https://github.com/google/heir/tree/main).
+"""
+
+from abc import ABC
+from typing import Annotated, Generic, TypeVar
+
+from xdsl.dialects.arith import signlessIntegerLike
+from xdsl.dialects.builtin import AnyIntegerAttr
+from xdsl.ir import Attribute, Dialect, Operation, SSAValue
+from xdsl.irdl import (
+    ConstraintVar,
+    IRDLOperation,
+    ParsePropInAttrDict,
+    irdl_op_definition,
+    operand_def,
+    prop_def,
+    result_def,
+)
+from xdsl.traits import Pure
+
+
+class ModArithOp(IRDLOperation, ABC):
+    pass
+
+
+_T = TypeVar("_T", bound=Attribute)
+
+
+class BinaryOp(ModArithOp, ABC, Generic[_T]):
+    """
+    Simple binary operation
+    """
+
+    T = Annotated[Attribute, ConstraintVar("T"), _T]
```

Comment:
Is there any future where the type isn't `signlessIntegerLike`? I'm doing some related work at the moment, and it would make it easier if we could specialize as part of this PR, something like this:
```suggestion
class BinaryOp(ModArithOp, ABC):
    """
    Simple binary operation
    """

    T = Annotated[Attribute, ConstraintVar("T"), signlessIntegerLike]
```